third_party/gpus/cuda_configure.bzl

```diff
@@ -158,20 +147,110 @@ def _cudnn_install_basedir(repository_ctx):
   return cudnn_install_path
 
 
-def _cuda_version(repository_ctx):
-  """Detects the cuda version."""
-  if _TF_CUDA_VERSION in repository_ctx.os.environ:
-    return repository_ctx.os.environ[_TF_CUDA_VERSION].strip()
+_PREFIXES = {
+    "Linux": "lib",
+    "Darwin": "lib",
+    "Windows": "",
+}
+
+_EXTENSIONS = {
+    "Linux": "so",
+    "Darwin": "dylib",
+    "Windows": "lib",
+}
+
+def _lib_version(lib_name, repository_ctx, cpu_value, base_dir):
+  """Returns the version of the library parsed from its file name.
+
+  This method attempts to resolve the library to its canonical file, which
+  contains the full version in the file name and then extracts the major and
+  minor numbers from the version. For example, if the cublas library is
+  resolved to libcublas.so.7.5.18, then this function returns "7.5".
+
+  Args:
+    lib_name: The name of the library, such as "cudart"
+    repository_ctx: The repository context.
+    cpu_value: The name of the host operating system.
+    base_dir: The base directory the library is installed under.
+
+  Returns:
+    String containing the major and minor version numbers separated by a
+    period.
+  """
+  lib = _find_cuda_lib(lib_name, repository_ctx, cpu_value, base_dir)
+  lib_path = lib.path
+  basename = repository_ctx.path(lib_path).basename
+
+  if not cpu_value in _EXTENSIONS.keys():
+    auto_configure_fail("Not supported CPU value %s" % cpu_value)
+  expected_basename = "%s%s" % (_PREFIXES[cpu_value], lib_name)
+  extension = _EXTENSIONS[cpu_value]
+  if basename == "%s.%s" % (expected_basename, extension):
```

Comment:
The situation on Windows is a little bit complicated.
But basically, every `dll` file has a corresponding `lib` file.
During linking time, we only need to link against the `lib` files, as long as the "dll" files are in PATH, the executable will run.
And the `lib` files are non-versioned, they are just like `cuda.lib`, `cudann.lib` without version number. But the `dll` files are versioned, they are like `cudart32_80.dll`, `cudart64_80.dll`, `cudann64_5.dll`. When loading libraries in [dso_loader.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/dso_loader.cc#L63), we are actually loading the dlls, so `GetCudaVersion()` and `GetCudnnVersion()` must return correct version number.
But the `lib` files specified in BUILD files have to be non-versioned.